File: 09/01/PageObjects/BasePage.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace PageObjects
{
    class BasePage
    {
        IWebDriver Driver;

        protected BasePage(IWebDriver Driver)
        {
            this.Driver = Driver;
        }
            
        protected void Visit(string Url)
        {
            Driver.Navigate().GoToUrl(Url);
        }

        protected IWebElement Find(By Locator)
        {
            return Driver.FindElement(Locator);
        }

        protected void Click(By Locator)
        {
            Find(Locator).Click();
        }

        protected void Type(By Locator, string InputText)
        {
            Find(Locator).SendKeys(InputText);
        }

        protected bool IsDisplayed(By Locator)
        {
            return Find(Locator).Displayed;
        }
    }
}

File: code-examples/ruby/09_base_page_object/01/base_page.rb
require 'selenium-webdriver'

class BasePage

  def initialize(driver)
    @driver = driver
  end

  def visit(url_path)
    @driver.get ENV['base_url'] + url_path
  end

  def find(locator)
    @driver.find_element locator
  end

  def type(text, locator)
    find(locator).send_keys text
  end

  def submit(locator)
    find(locator).submit
  end

  def is_displayed?(locator)
    find(locator).displayed?
  end

end
